stdlib/public/core/StringGuts.swift

```diff
@@ -248,7 +248,22 @@ extension _StringGuts {
   ) -> Int? {
     #if _runtime(_ObjC)
     // Currently, foreign  means NSString
-    return _cocoaStringCopyUTF8(_object.cocoaObject, into: mbp)
+    guard let res = _cocoaStringCopyUTF8(_object.cocoaObject, into: mbp) else {
+      // If the NSString contains invalid UTF8 (e.g. unpaired surrogates), we
+      // can get nil from cocoaStringCopyUTF8 in situations where a character by
+      // character loop would get something more useful like repaired contents
+      var ptr = mbp.baseAddress._unsafelyUnwrappedUnchecked
+      var numWritten = 0
+      for cu in String(self).utf8 {
+        guard numWritten < mbp.count else { return nil }
+        ptr.initialize(to: cu)
+        ptr += 1
+        numWritten += 1
+      }
+      
+      return numWritten
+    }
+    return res
```

Comment:
One benefit of early-returns and `guard` vs `if` is to simplify nesting and ease code reading. Unfortunately, I think this use of `guard` makes it harder to understand. Would it be simpler to say:

```swift
if let res = _cocoaStringCopyUTF8(_object.cocoaObject, into: mbp) {
  return res
}

// If the NSString ....
...
return numWritten
```

Reply:
Fair enough